gdx-jnigen/src/main/java/com/badlogic/gdx/jnigen/build/toolchains/BaseToolchain.java

```diff
@@ -0,0 +1,214 @@
+            sourceCollectedCppFiles = collectFiles(rootDir.toPath(), cppIncludes, cppExcludes);
+            logger.info("Collected CPP files {}", sourceCollectedCppFiles);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    private void cleanBuildDirectory () {
+
+        logger.info("Cleaning build directory {} - abs {}", buildDirectory, buildDirectory.getAbsolutePath());
+        logger.info("Cleaning libs directory {} - abs {}", libsDirectory, libsDirectory.getAbsolutePath());
+
+        logger.error("Not actually wiping directories yet");
+    }
+
+    private void createBuildDirectory () {
+        logger.info("Creating build directory and copying source includes");
+
+        buildDirectory.mkdirs();
+        if (!buildDirectory.exists()) {
+            logger.error("Failure making file {}", buildDirectory.getAbsolutePath());
+            throw new RuntimeException("Failure to create build dir");
+        }
+
+        try {
+            //copy the source files all over
+            buildCFiles = new ArrayList<>();
+            buildCppFiles = new ArrayList<>();
+            for (File collectedCFile : sourceCollectedCFiles) {
+                File targetFile = new File(buildDirectory, collectedCFile.getName());
+                buildCFiles.add(targetFile);
+                Files.copy(collectedCFile.toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            }
+            for (File collectedCppFile : sourceCollectedCppFiles) {
+                File targetFile = new File(buildDirectory, collectedCppFile.getName());
+                buildCppFiles.add(targetFile);
+                Files.copy(collectedCppFile.toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            }
+        } catch (IOException e) {
+            logger.error("Failure to copy files into build directory", e);
+            throw new RuntimeException("Failure to copy files into build directory");
+        }
+
+    }
+
+    public static List<File> collectFiles (Path rootDir, String[] includes, String[] excludes) throws IOException {
+        List<File> matchedFiles = new ArrayList<>();
+
+        Files.walkFileTree(rootDir, new FileVisitor<Path>() {
+            @Override
+            public FileVisitResult preVisitDirectory (Path dir, BasicFileAttributes attrs) {
+                for (String exclude : excludes) {
+                    if (dir.endsWith(exclude)) {
+                        return FileVisitResult.SKIP_SUBTREE;
+                    }
+                }
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFile (Path file, BasicFileAttributes attrs) {
+                String filePath = file.toString();
+                boolean include = false;
+                for (String includePattern : includes) {
+                    if (filePath.matches(includePattern.replace("**", ".+"))) {
```

Comment:
`path/*.cpp` is also a valid path, if you don't want to pick up recursivly in sub directories